core-primitives/nonce-cache/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(assert_matches)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use thiserror_sgx as thiserror;
+}
+
+#[cfg(feature = "std")]
+use std::sync::RwLockWriteGuard;
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLockWriteGuard as RwLockWriteGuard;
+
+use crate::error::Result;
+use lazy_static::lazy_static;
+use std::sync::Arc;
+
+pub use nonce_cache::NonceCache;
+
+lazy_static! {
+	/// Global instance of a nonce cache
+	///
+	/// Concurrent access is managed internally, using RW locks
+	pub static ref GLOBAL_NONCE_CACHE: Arc<NonceCache> = Default::default();
+}
+
+pub mod error;
+pub mod nonce_cache;
+
+pub type NonceValue = u32;
```

Comment:
NonceValue is an `u32`. I think there is an open issue where we want to connect this to another type definition `Index`?

Also I first tried to use `u64` or even `u128` to be more safe from overflows - but unfortunately deeper down in the extrinsics code, we use `u32`, so I'm forced to use `u32` here too.

Reply:
Yes, I created the issue yesterday because I noticed the hardcoded `u32` in PR #487: https://github.com/integritee-network/worker/issues/495. 

I also think we shouldnt define the value here but rather in the settings or wherever we have our node primitves defined. But lets leave this task to the issue itself. This is atleast one step towards that goal, thanks.